---
title: "Introduction to Plotting"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/Plotting_In_R/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style

#dougs edit


---
class: inverse, center, middle

# Plotting in R (Base graphics)
<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
##Introduction

```{r setup_basePlotting, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```


R has excellent graphics and plotting capabilities. They are mostly found in following three sources.
 + base graphics
 + the lattice package
 +  the ggplot2 package
 
Base R graphics uses a pen and paper model for plotting while Lattice and ggplot2 packages are built on the routines first used in grid graphics.

---
## A pen and paper model

- Once plot is produced, can only add more elements, cannot remove.
- Makes it hard to update.
- But faster than more complex plotting systems.



---

## Scatter and linne Charts

First we'll produce a very simple graph using the values in a numeric vector:

```{r treatmentVector_basePlotting}
treatment <- c(0.02,1.8, 17.5, 55,75.7, 80)
```

---
##Scatter Plot

Now we plot the treatment vector with default parameters

```{r defaultPlot_basePlotting,eval=FALSE}
plot(treatment)
```



```{r defaultPlotEval_basePlotting, echo=FALSE,fig.width=8,fig.height=4.5,dpi=300,out.width="1920px",height="1080px"}
plot(treatment)
```

---
Now, let's customise this plot a little. 

First we can plot treatment using points overlayed by a line. We control this with the **type** argument.


```{r plotType_basePlotting,eval=TRUE,fig.width=8,fig.height=3.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="o")
```

---
We have many type options available to us. 

To see a complete list we can use **?plot**

.pull-left[
```{r plotTypeSEval_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="l")
```
]
.pull-right[
```{r plotTypeS_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="p")
```
]
---

We can add additional labels to our plot's axis and main/sub-title too.

We add a title with **main** argument and or a sub-title with the **sub** argument.

```{r plotLabels_basePlotting,eval=TRUE,fig.width=8,fig.height=3.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, main="My Plot", sub="a plot")
```

---

We can customise our x and y axis label with the **xlab** and **ylab** arguments respectively.

```{r plotAsisLabels_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, xlab="Position", ylab="score")
```

---

We can control the orientation of labels on axis using **las** argument.

.pull-left[
```{r plotAsisLasL_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, las=1)
```
]
.pull-right[
```{r plotAsisLasR_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, las=2)
```
]

---

We can control the size of points in our plot using the **cex** parameter.

.pull-left[
```{r plotAsisCexL_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, cex=2)
```
]
.pull-right[
```{r plotAsisCexR_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, cex=0.5)
```
]

---

We can control the type of points in our plot using the **pch** parameter.

.pull-left[
```{r plotAsisPchL_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, pch=1)
```
]
.pull-right[
```{r plotAsisPchR_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, pch=20)
```
]


---

Similarly when plotting a line we control size with **lwd** parameter.

.pull-left[
```{r plotAsisLwdL_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="l",lwd=10)
```
]
.pull-right[
```{r plotAsisLwdR_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="l",lwd=0.5)
```
]

---

We can also control the type of line with **lty** parameter.

.pull-left[
```{r plotAsisLtyL_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="l",lty=1)
```
]
.pull-right[
```{r plotAsisLtyR_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="l",lty=2)
```
]

---


An important parameter we can control is colour. We can control colour or lines or points using the **col** argument.

.pull-left[
```{r plotColL_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="l",
     col="red")
```
]
.pull-right[
```{r plotColR_basePlotting,eval=TRUE,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="l",
     col="dodgerblue")
```
]

---

You can find an extensive list of R colours here.

[R colour Chart](http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf)

---


Review **?plot** and **?par** for complete list of options.


---

The **plot** function vector will accept two vectors to be plotted against each other.

```{r plotTwoVectors_basePlotting,eval=TRUE,collapse=TRUE,fig.width=8,fig.height=3.5,dpi=300,out.width="1000px",height="1080px"}
control <- c(0, 20, 40, 60, 80,100)
plot(treatment,control)
```

---

But we often want mulitple lines in same plot. So if we want to plot scores for control and treatment against position we will need a new method.

We can add an addition line to our existing plot using the **lines()** function.

```{r lines_basePlotting,eval=TRUE,collapse=TRUE,fig.width=8,fig.height=3.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="o", col="blue")
lines(control, type="o", pch=22, lty=2, col="red")
```

---
But the new line doesnt quite fit into our original plot.


We can extend our x or y axis by specifying values to **xlim** **ylim** arguments.

```{r yxlim_basePlotting,eval=TRUE,collapse=TRUE,fig.width=8,fig.height=3.5,dpi=300,out.width="1000px",height="1080px"}
control <- c(0, 20, 40, 60, 80,100)
plot(treatment, type="o", col="blue",ylim=c(0,100))
lines(control, type="o", pch=22, lty=2, col="red")
```

---

Next let's change the axes labels to match our data and add a legend. We'll also compute the y-axis values using the **range** function 
so any changes to our data will be automatically reflected in our graph. 

range() returns a vector containing the minimum and maximum of all the given arguments.


Calculate range from 0 to max value of data
```{r range_basePlotting}
g_range <- range(0, treatment, control)
g_range
```


---

Plot treatment using y axis that ranges from 0 to max value in treatment or control vector.  Turn off axes and annotations (axis labels) so we can specify them ourselves. We turn of axis and annotation plotting using **axes=FALSE** and **ann=FALSE**

```{r plotNoAxesNoAnnotation_basePlotting,fig.width=8,fig.height=4.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="o", col="blue", 
     ylim=g_range,axes=FALSE, ann=FALSE)
```

---

We can now create our own X axis by using the **axis()** function. We specify the **side** argument for where to place axis, the **at** argument to specify where to put axis ticks and **lab** argument to specify labels for axis ticks.

```{r axis_basePlotting,eval=FALSE}
axis(side=1, at=1:6, lab=c("Mon","Tue","Wed","Thu","Fri","Sat"))
```

```{r axisEval_basePlotting,eval=TRUE,echo=FALSE,fig.width=8,fig.height=4,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="o", col="blue", 
     ylim=g_range,axes=FALSE, ann=FALSE)
axis(1, at=1:6, lab=c("Mon","Tue","Wed","Thu","Fri","Sat"))

```

---

We can make our y axis with horizontal labels that display ticks at every 20 marks in a similar way.

We specify our **side** and use **rep()** function to make axis tick postions for **at** argument.

```{r axisSides_basePlotting,eval=FALSE}
axis(2, las=1, at=rep(0,g_range[2],by=20))
```

```{r axisSidesEval_basePlotting,eval=TRUE,echo=FALSE,fig.height=3.5,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="o", col="blue", 
     ylim=g_range,axes=FALSE, ann=FALSE)
axis(1, at=1:6, lab=c("Mon","Tue","Wed","Thu","Fri","Sat"))
axis(2, las=1, at=20*0:g_range[2])

```

---

We can now add a box around our plot using the **box()** function.
```{r box_basePlotting,eval=FALSE}
box()
```

```{r boxEval_basePlotting,eval=TRUE,echo=FALSE,fig.height=4,dpi=300,out.width="1000px",height="1080px"}
plot(treatment, type="o", col="blue", 
     ylim=g_range,axes=FALSE, ann=FALSE)
axis(1, at=1:6, lab=c("Mon","Tue","Wed","Thu","Fri","Sat"))
axis(2, las=1, at=20*0:g_range[2])
box()
```


---

Now i can add my control data using lines argument.

```{r addLines_basePlotting,eval=FALSE,echo=TRUE,collapse=TRUE}
lines(control, type="o", pch=22, lty=2, col="red")

```

```{r addLinesEval_basePlotting,eval=TRUE,echo=FALSE,fig.width=8,fig.height=3.5,dpi=300,out.width="1920px",height="1080px"}
plot(treatment, type="o", col="blue", 
     ylim=g_range,axes=FALSE, ann=FALSE)
axis(1, at=1:6, lab=c("Mon","Tue","Wed","Thu","Fri","Sat"))
axis(2, las=1, at=20*0:g_range[2])
box()
lines(control, type="o", pch=22, lty=2, col="red")
```
---

Finally we may wish to add a legend to out plot. We can add a legend to current plot using the **legend()** function.

We need to specify where to place legend in plot, the names in legend to **legend** argument and any additional point/line type configuration we used, e.g blue/red.

```{r legend_basePlotting,eval=FALSE}
legend("topleft",legend=c("treatment","control"),
       col=c("blue","red"), pch=21:22, lty=1:2);  
```

```{r legendEval_basePlotting,echo=FALSE,fig.width=8,fig.height=3.5,dpi=300,out.width="1920px",height="1080px"}
plot(treatment, type="o", col="blue", ylim=g_range,axes=FALSE, ann=FALSE)
axis(1, at=1:6, lab=c("Mon","Tue","Wed","Thu","Fri","Sat"))
axis(2, las=1, at=20*0:g_range[2])
box()

lines(control, type="o", pch=22, lty=2, col="red")
legend("topleft",legend=c("treatment","control"),
       col=c("blue","red"), pch=21:22, lty=1:2);  
```
---

##Bar Charts	
	
	
Base graphics has a useful built in function for bar charts. The **barplot()** function.
We can simply pass our numeric vector to this function to get our barchart.

```{r barplot_basePlotting,eval=FALSE}
barplot(treatment)
```

```{r barplotEval_basePlotting,echo=FALSE,fig.width=8,fig.height=4,dpi=300,out.width="1920px",height="1080px"}
barplot(treatment)
```

---

The **barplot()** function hasn't added any labels by default. We can speciy our own however using the **names.arg** argument. **names.arg**  is a vector of names to be plotted below each bar or group of bars. 


```{r barplotNames_basePlotting,eval=TRUE,fig.width=8,fig.height=3.5,dpi=300,out.width="1920px",height="1080px"}
barplot(treatment,
        names.arg=c("Mon","Tue","Wed","Thu","Fri","Sat"))
```

---

If my vector was named however, then my vectors names would be used for labels. We use **names()** function to add names to our vector then we replot.


```{r barplotPreNames_basePlotting,eval=TRUE,fig.width=8,fig.height=3.5,dpi=300,out.width="1920px",height="1080px"}
names(treatment) <- c("Mon","Tue","Wed","Thu","Fri","Sat")
barplot(treatment)
```


---
 
Let's now read the data from the example.txt data file.  

Read values from tab-delimited example.txt
```{r readTable_basePlotting}
data <- read.table("data/example.txt", header=T, sep="\t")
```



---
We can now plot data from a matrix with a stacked barchart. Note we use **as.matrix()** function to convert our data.frame to a matrix.


```{r barplotMatrix_basePlotting,eval=TRUE,fig.width=8,fig.height=4,dpi=300,out.width="1920px",height="1080px"}
barplot(as.matrix(data))
```

---
Now we can plot data from a matrix with side-by-side barchart using the **beside** argument


```{r barplotMatrixBeside_basePlotting,eval=TRUE,fig.width=8,fig.height=4,dpi=300,out.width="1920px",height="1080px"}
barplot(as.matrix(data),beside=TRUE)
```

---
Now we can put it together and plot data with adjacent bars using colors and place the legend at the top-left corner.


```{r barplotComplex_basePlotting,eval=FALSE}
barplot(as.matrix(data), main="Data", ylab= "Total", beside=TRUE, 
        col= c("lightblue", "mistyrose", "lightcyan","lavender",
               "cornsilk","maroon"))
legend("topleft", c("Mon","Tue","Wed","Thu","Fri","Sat"), cex=0.8,
        fill=  c("lightblue", "mistyrose", "lightcyan",
                         "lavender","cornsilk","maroon"));
```


```{r barplotComplexEval_basePlotting,echo=FALSE,fig.width=8,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
barplot(as.matrix(data), main="Data", ylab= "Total", beside=TRUE, col= c("lightblue", "mistyrose", "lightcyan","lavender", "cornsilk","maroon"))
legend("topleft", c("Mon","Tue","Wed","Thu","Fri","Sat"), cex=0.8,
fill=  c("lightblue", "mistyrose", "lightcyan","lavender", "cornsilk","maroon"));

```

---

##Histograms


Base graphics has a useful built in function for histograms too. The **histogram()** function.
We can simply pass our numeric vector to this function to get our barchart.

```{r hist_basePlotting,echo=TRUE,fig.width=8,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
hist(treatment)  
```


---

Similar cutomisation exists as for other plots.
```{r histColourYlim_basePlotting,eval=FALSE}
hist(treatment, col="lightblue", ylim=c(0,10))
```

```{r histColourYlimEval_basePlotting,echo=FALSE,fig.width=5,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
hist(treatment, col="lightblue", ylim=c(0,10),cex=0.5,cex.main=0.8)
```
---

We can create more fine grained histogram by specify the number of required bins to the **breaks** argument.
```{r histBreaks_basePlotting,eval=FALSE}
hist(treatment, col="lightblue",breaks = 2)
```

```{r histBreaksEval_basePlotting,echo=FALSE,fig.width=5,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
hist(treatment, col="lightblue", ylim=c(0,10),cex=0.5,cex.main=0.8,breaks = 2)
```

---

##Dot charts

Base graphics has a useful built in function for dotcharts too. The **dotchart()** function.
We can simply pass our numeric vector to this function to get our barchart.

Here we use the function **t** to return the transpose of a matrix.
```{r dotchart_basePlotting,eval=FALSE}
dotchart(t(data))	
```
```{r dotchartEval_basePlotting,echo=FALSE,fig.width=8,fig.height=4,dpi=300,out.width="1920px",height="1080px"}
dotchart(t(data))  
```

---

Let's make the dotchart a little more colourful: 

Now we create a coloured dotchart for autos with smaller labels
```{r dotchartCol_basePlotting,eval=FALSE}
dotchart(t(data), color=c("red","blue"),
         main="Dotchart", cex=0.8)	
```
```{r dotchartColEval_basePlotting,echo=FALSE,fig.width=6,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
dotchart(t(data), color=c("red","blue"),main="Dotchart", cex=0.8)  
```

---

##Box plots

The final plot we will look at is a box and whisker plot. 

Boxplots allow you to quickly review data distributions, showing the median and 1st/3rd quartile.

![](imgs/box_whisker_describe.png)

---

First lets read in the gene expression data
```{r readForBoxplot_basePlotting,eval=T,fig.width=6,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
exprs <- read.delim("data/gene_data.txt",sep="\t",h=T,row.names = 1)
head(exprs)
```

---

Now we can use the **boxplot()** function on our data.frame to get our boxplot
```{r Boxplot_basePlotting,eval=T,fig.width=6,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
boxplot(exprs)
```

---

Perhaps it would look better on a log scale. We can add addition colours and labels as with other plots.

```{r BoxplotEval_basePlotting,eval=T,fig.width=6,fig.height=3,dpi=300,out.width="1920px",height="1080px"}
boxplot(log2(exprs),ylab="log2 Expression",
        col=c("red","red","blue","blue"))
```

---
##Combining Plots

R makes it easy to combine multiple plots into one overall graph, using either the **par( )** or **layout( )** function. 
With the **par( )** function, you can include the option mfrow=c(nrows, ncols) to create a matrix of nrows x ncols plots that are filled in by row.
mfcol=c(nrows, ncols) fills in the matrix by columns.

Define a layout with 2 rows and 2 columns

```{r parMfrow_basePlotting,eval=FALSE}
par(mfrow=c(2,2))
```

---

Here, we will use different dataset with two columns each for treated and untreated samples.

```{r readForParMfrow_basePlotting}
data1 <- read.table("data/gene_data.txt", header=T, sep="\t")
head(data1)
```
---

Plot histograms for different columns in the data frame separately. This is not very efficient. 
You could also do it more efficiently using for loop.

```{r setParMfrow_basePlotting,eval=FALSE}
par(mfrow=c(2,2))
hist(data1$Untreated1)
hist(data1$Treated2)
hist(data1$Untreated2)
boxplot(data1$Treated1)
```

```{r setParMfrowEval_basePlotting,echo=FALSE,fig.width=8,fig.height=5,dpi=300,out.width="490px",height="270px"}
par(mfrow=c(2,2))
hist(data1$Untreated1)
hist(data1$Treated2)
hist(data1$Untreated2)
boxplot(data1$Treated1)
```

---
##Saving your plots



There are many different ways of saving your plots in R. 

The only argument you would need is name of file in which you want to save the plot.

Plotting commands then can be entered as usual.
The output would be redirected to the file. 

When you're done with your plotting commands, enter the dev.off() command. 

```{r save_basePlotting,eval=FALSE}
bmp(filename, width = 480, height = 480, units = "px", 
    point-size = 12)
jpeg(filename, width = 480, height = 480, units = "px", 
     point-size = 12, quality = 75)
```
---

Saving in bitmap format
```{r save_bmp_basePlotting,eval=FALSE}
bmp(file = "control.bmp")
plot(control)
dev.off()
```
---
Saving in jpeg format
```{r save_jpeg_basePlotting,eval=FALSE}
jpeg(file = "control.jpg", quality = 20)
plot(control)
dev.off()
```

---

Saving in postscript format

```{r save_ps_basePlotting,eval=FALSE}
postscript(file = "control.ps")
plot(control)
dev.off()
```

---
saving in pdf format

```{r save_pdf_basePlotting,eval=FALSE}

pdf(file = "control.pdf", paper = "A4")
plot(control)
dev.off()
```


---

Exercise on base plotting can be found [here](../../exercises/Plotting_exercise.html)

---

Answers for baseplotting can be found [here](../../answers/Plotting_answers.html)

---
